platform/logger.cpp

```
@@ -0,0 +1,273 @@
+static uint32_t log_count_ = 0;
+#endif
+
+// Globals related to ISR logging
+static LOG_DATA_TYPE_ log_isr_str_[LOG_SINGLE_STR_SIZE_];
+static bool log_isr_buf_busy = false;
+static EventQueue *log_isr_equeue_;
+
+// Globals related to time printing
+static const ticker_data_t *const log_ticker_ = get_us_ticker_data();
+static bool time_enable_ = true;
+
+// Globals related to external buffer capturing
+static LOG_DATA_TYPE_ *extern_buf_ = NULL;
+static uint32_t xbuf_count_ = 0;
+
+static void log_update_buf(const char *data, int32_t size)
+{
+    if ((NULL == extern_buf_) || (NULL == data) || (size <= 0)) {
+        return;
+    }
+    uint32_t count = 0;
+    while(size--)
+    {
+        extern_buf_[xbuf_count_++] = data[count++];
+        xbuf_count_ %= MBED_CONF_EXTERNAL_BUFFER_SIZE;
+    }
+}
+
+#if defined (MBED_ID_BASED_TRACING)
+// Note: ISR data is lossy, and with too many interrupts it might be
+// over-written and last valid ISR data will be printed.
+static void log_isr_id_data()
+{
+    mbed_log_lock();
+    log_isr_buf_busy = true;
+    uint32_t count = 0;
+    while(log_count_--) {
+        fprintf(stderr, "0x%x ", log_isr_str_[count++]);
+    }
+    log_isr_buf_busy = false;
+    mbed_log_unlock();
+}
+#else
+// Note: ISR data is lossy, and with too many interrupts it might be
+// over-written and last valid ISR data will be printed.
+static void log_isr_queue()
+{
+    mbed_log_lock();
+    log_isr_buf_busy = true;
+    fputs(log_isr_str_, stderr);
+    log_isr_buf_busy = false;
+    mbed_log_unlock();
+}
+
+static void log_str_isr_data(const char *format, va_list args)
+{
+    if (true == log_isr_buf_busy) {
+        return;
+    }
+    core_util_critical_section_enter();
+    volatile uint64_t time = ticker_read_us(log_ticker_);
+    int32_t count = 0;
+    if (time_enable_) {
+        count = snprintf(log_isr_str_, LOG_SINGLE_STR_SIZE_, "[%-8lld]", time);
+        if (count < 0 || count > LOG_SINGLE_STR_SIZE_) {
+            count = 0;
```

maybe we should not continue printing in case of fail to avoid unexpected behavior..